### PR TITLE
Scoring pattern refactor

### PR DESCRIPTION
**Pull request by halfak**
* New ScoringSystem replaces ScoreProcessor pattern
* Scoring multiple models for the same revision is essentially free[1]
* Base feature map production is cleaner
* Increments version to 0.8.0 to reflect some major changes
* routes for v1 and v2 stay exactly the same

1. ...depending on feature dependecy tree overlap.

### CODE REVIEW

**Comment by halfak** [comment]
So, when I rebuilt this PR, I explicitly moved score_processor.py to scoring_system.py and then did the changes, but still the file-diff shows a complete removal and re-addition of the file.  I don't think there's anything that can be done :\

File: ores/scoring_context.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+logger = logging.getLogger(__name__)
+
+
+class ScoringContext(dict):
+
+    def __init__(self, name, model_map, extractor):
+        """
+        :Parameters:
+            name : str
+               The  that this scorer is being used for
```
Comment (schana) [inline ores/scoring_context.py]: Unclear

File: ores/scoring_context.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            scorer_models : dict
+               A mapping between names and
+               :class:`revscoring.ScorerModel`
+               instances
+            extractor : :class:`revscoring.Extractor`
+               An extractor to use for gathering feature values
+        """
+        super().__init__()
+        self.name = str(name)
```
Comment (schana) [inline ores/scoring_context.py]: This should be unnecessary with the contract specified by the docstring

File: ores/scoring_context.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+
+class ScoringContext(dict):
+
+    def __init__(self, name, model_map, extractor):
+        """
+        :Parameters:
+            name : str
+               The  that this scorer is being used for
+            scorer_models : dict
```
Comment (schana) [inline ores/scoring_context.py]: needs updated to 'model_map'

File: ores/scoring_context.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+               A mapping between names and
+               :class:`revscoring.ScorerModel`
+               instances
+            extractor : :class:`revscoring.Extractor`
+               An extractor to use for gathering feature values
+        """
+        super().__init__()
+        self.name = str(name)
+        self.update(model_map)
```
Comment (schana) [inline ores/scoring_context.py]: I think it would make more sense to maintain this dictionary separately instead of absorbing it into `self`

File: ores/scoring_context.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            if include_features:
+                base_feature_map = self.solve_base_feature_map(
+                    model_name, dependency_cache=root_cache)
+                model_scores[model_name]['features'] = base_feature_map
+
+        return model_scores
+
+    def solve_features(self, model_name, dependency_cache=None):
+        """
+        """
```
Comment (schana) [inline ores/scoring_context.py]: Why the empty docstring?

File: ores/scoring_context.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        return {str(f): v
+                for f, v in zip(features, feature_values)}
+
+    def process_score(self, model_name, dependency_cache=None):
+        """
+        """
+        version = self[model_name].version
+
+        start = time.time()
+        feature_values = list(self.solve_features(
```
Comment (schana) [inline ores/scoring_context.py]: unnecessary cast

File: ores/scoring_context.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        start = time.time()
+        score = self[model_name].score(feature_values)
+        logger.debug("Scored features for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        return score
```
Comment (schana) [inline ores/scoring_context.py]: The timing logic pollutes what the function is trying to accomplish

File: ores/scoring_context.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    def _generate_root_datasources(self, model_names):
+        for model_name in model_names:
+            for dependency in dependencies.dig(self.model_features(model_name)):
+                if isinstance(dependency, Datasource):
```
Comment (schana) [inline ores/scoring_context.py]: Why is this type checking necessary?

File: ores/scoring_context.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+                The names of a :class:`~revscoring.ScorerModel` to
+                extract the roots dependencies for
+        """
+        # Make a copy of dependency_caches
+        root_caches = {}
+        for rev_id in rev_ids:
+            injection_cache = injection_caches.get(rev_id) \
+                              if injection_caches is not None else {}
+            root_caches[rev_id] = {key: value
+                                   for key, value in injection_cache.items()}
```
Comment (schana) [inline ores/scoring_context.py]: Could the `copy` library be used here instead?

File: ores/scoring_context.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+        # Check each extraction for errors
+        errors = {}
+        for rev_id, (error, root_vals) in zip(rev_ids, error_root_vals):
+            if error is not None:
+                errors[rev_id] = error
+                del root_caches[rev_id]
+        logger.debug("Extracted root datasources for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, set(model_names), rev_ids,
+                             round(time.time() - start, 3)))
```
Comment (schana) [inline ores/scoring_context.py]: `time.time()` is guaranteed to be that precise

File: ores/scoring_context.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            injection_cache = injection_caches.get(rev_id) \
+                              if injection_caches is not None else {}
+            root_caches[rev_id] = {key: value
+                                   for key, value in injection_cache.items()}
+
+        # Find our root datasources
+        root_datasources = \
+            list(set(self._generate_root_datasources(model_names)))
+
+        # Extract the root datasource and update the root_caches in-place
```
Comment (schana) [inline ores/scoring_context.py]: the side-effect should be more explicitly described through something other than a comment in the caller's code

File: ores/scoring_context.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        start = time.time()
+        error_root_vals = self.extractor.extract(
+            rev_ids, root_datasources, caches=root_caches)
+
+        # Check each extraction for errors
+        errors = {}
+        for rev_id, (error, root_vals) in zip(rev_ids, error_root_vals):
```
Comment (schana) [inline ores/scoring_context.py]: `root_vals` isn't used and could be replaced with `_` to avoid mistaking `root_caches` for it

File: ores/scoring_context.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+                    extractor: ptwiki
+
+            extractors:
+                enwiki_api: ...
+                ptwiki_api: ...
+
+            scorer_models:
+                enwiki_damaging_2014: ...
+                enwiki_good-faith_2014: ...
+        """
```
Comment (schana) [inline ores/scoring_context.py]: Is there a formal grammar for specifying how the configuration should be? That would be more useful IMO than an (outdated?) example

File: ores/scoring_systems/celery_queue.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+logger = logging.getLogger(__name__)
+
+APPLICATIONS = []
+
+DEFAULT_CELERY_QUEUE = "celery"
+
+
+@before_task_publish.connect
+def update_sent_state(sender=None, body=None, **kwargs):
+
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: unnecessary newline

File: ores/scoring_systems/celery_queue.py [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+import mwapi.errors
+import revscoring.errors
+from celery.signals import before_task_publish
+
+from ..errors import TimeoutError, ScoreProcessorOverloaded
+from .scoring_system import ScoringSystem
+
+logger = logging.getLogger(__name__)
+
+APPLICATIONS = []
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: All caps generally denotes a static resource; this seems to not be the case here

File: ores/scoring_systems/celery_queue.py [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+        backend = task.backend if task else application.backend
+
+        logger.debug("Setting state to 'SENT' for {0}".format(body['id']))
+        backend.store_result(body['id'], result=None, status="SENT")
+
+
+class CeleryQueue(ScoringSystem):
+
+    def __init__(self, *args, application, queue_maxsize=None, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: I find this pattern hard to follow and un-pythonic

File: ores/scoring_systems/celery_queue.py [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+        self.application = application
+        self.queue_maxsize = int(queue_maxsize) if queue_maxsize is not None \
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: It seems like `queue_maxsize` is necessary (L211). Also, it should be the caller's responsibility to give a valid value instead of casting it here.

File: ores/scoring_systems/celery_queue.py [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+                raise
+            logger.info("Completed generating score map for" +
+                        "{0}:{1}:{2}"
+                        .format(context_name, set(model_names), rev_id))
+            return score_map
+
+        APPLICATIONS.append(application)
+
+        self._process_score_map = _process_score_map
+        self._lookup_score_in_map = _lookup_score_in_map
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: There's a lot going on in `__init__` that could be clarified by splitting it out into separate functions

File: ores/scoring_systems/celery_queue.py [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+        if self.queue_maxsize is not None and self.redis is None:
+            logger.warning("No redis connection.  Can't check queue size")
+
+        expected_errors = (revscoring.errors.RevisionNotFound,
+                           revscoring.errors.PageNotFound,
+                           revscoring.errors.UserNotFound,
+                           revscoring.errors.DependencyError,
+                           mwapi.errors.RequestError,
+                           mwapi.errors.TimeoutError,
+                           TimeoutError)
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: Since there's a name collision already, it would be helpful if the second `TimeoutError` was fully qualified

File: ores/scoring_systems/celery_queue.py [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+                score_map = score_map_result.get(timeout=self.timeout)
+            except Exception as e:
+                logger.error(e)
+                raise
+            logger.info("Found up {0} in {1}!".format(model_name, result_id))
+            return score_map[model_name]
+
+        @self.application.task(throws=expected_errors,
+                               queue=DEFAULT_CELERY_QUEUE,
+                               priority=0)
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: What is the priority used to determine for these two tasks?

File: ores/scoring_systems/celery_queue.py [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+Matches <password>@<host>:<port>
+"""
+
+
+def redis_from_url(url):
+    """
+    Converts a redis URL used by celery into a `redis.Redis` object.
+    """
+    # Makes sure that we only try to import redis when we need
+    # to use it
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: Is it ever not used? This seems like a good candidate to split out, since it's a different level of abstraction that could benefit from polymorphism

File: ores/scoring_systems/celery_queue.py [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+                raise
+            logger.info("Found up {0} in {1}!".format(model_name, result_id))
+            return score_map[model_name]
+
+        @self.application.task(throws=expected_errors,
+                               queue=DEFAULT_CELERY_QUEUE,
+                               priority=0)
+        def _process_score_map(
+             context_name, model_names, rev_id, root_cache=None,
+             injection_cache=None, include_features=False):
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: increasing this indentation would help readability

File: ores/scoring_systems/celery_queue.py [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+                for model_name in missing_models:
+                    task_id = context.format_id_string(
+                        model_name, rev_id, injection_cache=injection_cache)
+                    score_result = self._lookup_score_in_map.apply_async(
+                        args=(result.id, model_name), task_id=task_id,
+                        expires=self.timeout)
+                    if rev_id in results:
+                        results[rev_id][model_name] = score_result
+                    else:
+                        results[rev_id] = {model_name: score_result}
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: Breaking this big `for` loop into its smaller parts would help readability

File: ores/scoring_systems/celery_queue.py [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+        context = self[context_name]
+
+        inprogress_results = inprogress_results or {}
+
+        # Generate score results
+        results = {}
+        for missing_models, rev_ids in missing_model_set_revs.items():
+            for rev_id in rev_ids:
+                injection_cache = injection_caches.get(rev_id) \
+                                  if injection_caches is not None else None
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: This clause is repeated many times. What is the reason for this complexity / flexibility?

File: ores/scoring_context.py (hunk 2 above)
Comment (halfak) [inline ores/scoring_context.py]: Cheap runtime type checking like this is nice for catching issues.  I like str() and int() because they are free in the case that they encounter the right type.  

```
$ python
Python 3.5.1+ (default, Mar 30 2016, 22:46:26) 
[GCC 5.3.1 20160330] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> foo = "bar"
>>> id(foo)
139939565130224
>>> id(str(foo))
139939565130224
```

File: ores/scoring_context.py (hunk 4 above)
Comment (halfak) [inline ores/scoring_context.py]: It's nice to reference the models contained by their names with the __getitem__, __setitem__ syntax.  We do this with contexts in the ScoringSystem too.  I'm not married to this pattern, but it's served us well so far.

File: ores/scoring_context.py (hunk 5 above)
Comment (halfak) [inline ores/scoring_context.py]: Good Q.  Will fix in update

File: ores/scoring_context.py (hunk 7 above)
Comment (halfak) [inline ores/scoring_context.py]: What do you suggest?

File: ores/scoring_context.py (hunk 8 above)
Comment (halfak) [inline ores/scoring_context.py]: As the function name implies, we'd like to have the root Datasources.  There are other items that can be found at the root of a dependency tree -- like Constant.

File: ores/scoring_context.py (hunk 9 above)
Comment (halfak) [inline ores/scoring_context.py]: Actually, now that I look at it, a simple `dict()` wrapping will work nicely. 

```
$ python
Python 3.5.1+ (default, Mar 30 2016, 22:46:26) 
[GCC 5.3.1 20160330] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> d = {1: "foo", 2: "bar"}
>>> id(d)
139893406233160
>>> id(dict(d.items()))
139893379005128
```

File: ores/scoring_context.py (hunk 10 above)
Comment (halfak) [inline ores/scoring_context.py]: I'm not sure it needs to be all that precise here.

File: ores/scoring_context.py (hunk 11 above)
Comment (halfak) [inline ores/scoring_context.py]: What do you suggest?

File: ores/scoring_context.py (hunk 13 above)
Comment (halfak) [inline ores/scoring_context.py]: The example looks up to date to me.  The configuration is in yaml.

File: ores/scoring_systems/celery_queue.py (hunk 16 above)
Comment (halfak) [inline ores/scoring_systems/celery_queue.py]: Citation needed?

File: ores/scoring_systems/celery_queue.py (hunk 21 above)
Comment (halfak) [inline ores/scoring_systems/celery_queue.py]: Yes.  You do not need to use redis as a backend.

File: ores/scoring_systems/celery_queue.py (hunk 23 above)
Comment (halfak) [inline ores/scoring_systems/celery_queue.py]: It's two function calls and a conditional.  How would you break it up?

File: ores/scoring_systems/celery_queue.py (hunk 24 above)
Comment (halfak) [inline ores/scoring_systems/celery_queue.py]: It handles the possibility that injection_caches is not specified.  It's a one-liner with variable names that are a bit too long.   

I don't see that this is very complex, but I guess the flexibility is in place to simplify the call pattern.

File: ores/scoring_systems/celery_queue.py (hunk 16 above)
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: PEP 20 ;) It basically boils down to extensibility vs readability. This method will also lead to runtime errors as opposed to something that can be caught by IDEs/analyzers.

File: ores/scoring_systems/celery_queue.py [hunk 25]
```diff
@@ -0,0 +1,287 @@
...
+                if rev_id in rev_scores and \
+                   model_name in rev_scores[rev_id]:
+                    continue
+
+                task_id = context.format_id_string(
+                    model_name, rev_id, injection_cache=injection_cache)
+                score_result = \
+                    self._lookup_score_in_map.AsyncResult(task_id)
+                if score_result.state in ("REQUESTED", "SENT", "STARTED",
+                                          "SUCCESS"):
```
Comment (schana) [inline ores/scoring_systems/celery_queue.py]: These states should be defined as constants (for `REQUESTED` and `SENT`) and referenced from `celery.states` for the others

File: ores/scoring_systems/celery_queue.py (hunk 16 above)
Comment (halfak) [inline ores/scoring_systems/celery_queue.py]: It looks like this is a place where we disagree.  It seems to me that python's system of flexible arguments was designed for purposes like this.  In fact, my analyzer (language-python in Atom) works just fine with these argument patterns.  Because I do not explicitly capture the arguments that are not specified, I do not have access to them in this scope and that is exactly how it should be.  

What is the real readability issue here?  It seems to me that it is trivial to read this as "There's a set of arguments and key-word-arguments that are passed directly to ScoringSystem.__init__.  See ScoringSystem.__init__ for details.

**Comment by Ladsgroup** [comment]
First I wanted to test this out in ores-experiment.
1- Lots of changes need to be done in ores-wikimedia-config and ores-wmflabs-config
2-
Celery fails like this:
[2016-07-19 15:34:06,495: ERROR/MainProcess] consumer: Cannot connect to amqp://guest:**@127.0.0.1:5672//: [Errno 111] Connection refused.
Trying again in 2.00 seconds...

(Maybe my config changes were not enough. I put them [[https://gist.github.com/Ladsgroup/d66eec0c6857089dc0a23b40e56f8918 | here]])
3- Web when you try to reach a score fails like this:
```
10.68.21.68 - - [19/Jul/2016 15:16:31] "GET /v2/scores/enwiki/?models=damaging|goodfaith&revids=729078817 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/srv/ores/venv/lib/python3.4/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/srv/ores/config/submodules/ores/ores/wsgi/routes/v2/scores.py", line 105, in score_model_revisions_v2
    scores_doc['scores'][rev_id][model_name]['score']
KeyError: 'score'
```

**Comment by halfak** [comment]
Issue fixed with last commit.  Try this again: http://ores-experiment.wmflabs.org/v2/scores/enwiki/?models=damaging|goodfaith&revids=729078817
